src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+    internal static partial class WaitSubsystem
+    {
+        [EagerStaticClassConstruction] // used during lazy class construction
+        private static class HandleManager
+        {
+            private const int InitialCapacity = 4;
+
+            /// <summary>
+            /// Prevents reusing a deleted index until at least there are this many deleted indexes in the free list. The
+            /// purpose is to allow free handles to remain free for some time so that we can discover invalid attempts at using
+            /// free handles.
+            /// </summary>
+            private const int MinimumFreeIndexCountBeforeReuse = 64;
+
+            private static WaitableObject[] s_objects = new WaitableObject[InitialCapacity];
+            private static int s_count;
+            private static LowLevelQueue<int> s_freeIndexes;
```

Comment:
Maybe something like this could work:

```
int s_deletedHandleHeadIndex;
int s_deletedHandleTailIndex;
...

    assert(s_handles[s_deletedHandleTailIndex] == -1); // end of queue
    s_handles[s_deletedHandleTailIndex] = deletedIndex;
    s_handles[deletedIndex] = -1;
    s_deletedHandleTailIndex = deletedIndex;
```

Then it can maintain queue-style info with your data structure. I'll give it a try.